cp ../Step4_fix_het_TE_region/Fix_region.NewRef.fasta ./
perl ~/BigData/00.RD/Genome_Submission/fasta2apg.pl -i Fix_region.NewRef.fasta -o Fix_region.NewRef.fasta.gap
#remove and mosk orlando sequence from genome
python revision_FC.py
#make a copy to save a backup of new agp file
cp PGA.masked.agp PGA.masked.raw.agp
#edit gap lines generated by revision_FC.py, remove extra gap lines, save a backup
cp PGA.masked.agp PGA.masked.edit_backup.agp
#change U to N for 100 bp gaps
sed 's/U/N/g' Fix_region.NewRef.fasta.gap/PGA.masked.agp > PGA.masked.agp
#create fasta from agp to test if the file is correct: correct
perl agp2fasta.pl PGA.masked.agp PGA.masked.contig.fa
perl ~/BigData/00.RD/Genome_Submission/fasta2apg.pl -i supercontigs.fa -o supercontigs.fa.gap

#change gaps over 100kb or smaller than 100bp into 100bp
python revision_FC_fix_gaps.py --input supercontigs.fa
perl ~/BigData/00.RD/Genome_Submission/fasta2apg.pl -i supercontigs.fa.fix_gaps.fa -o supercontigs.fa.fix_gaps.fa.gap
sed 's/PGA_scaffold/scaffold/' supercontigs.fa.fix_gaps.fa > supercontigs.fa.fix_gaps.final.fa

